src/backend/distributed/utils/shard_pruning.c

```diff
@@ -0,0 +1,1245 @@
+				break;
+			}
+		}
+	}
+
+	return containsFalseClause;
+}
+
+
+/*
+ * FastShardPruning is a higher level API for FindShardInterval function. Given the
+ * relationId of the distributed table and partitionValue, FastShardPruning function finds
+ * the corresponding shard interval that the partitionValue should be in. FastShardPruning
+ * returns NULL if no ShardIntervals exist for the given partitionValue.
+ *
+ * FIXME: routine will be deleted after #1353 is merged.
+ */
+ShardInterval *
+FastShardPruning(DistTableCacheEntry *cacheEntry, Datum partitionValue)
+{
+	int shardCount = cacheEntry->shardIntervalArrayLength;
+	ShardInterval **sortedShardIntervalArray = cacheEntry->sortedShardIntervalArray;
+	bool useBinarySearch = false;
+	char partitionMethod = cacheEntry->partitionMethod;
+	FmgrInfo *shardIntervalCompareFunction = cacheEntry->shardIntervalCompareFunction;
+	bool hasUniformHashDistribution = cacheEntry->hasUniformHashDistribution;
+	FmgrInfo *hashFunction = NULL;
+	ShardInterval *shardInterval = NULL;
+
+	/* determine whether to use binary search */
+	if (partitionMethod != DISTRIBUTE_BY_HASH || !hasUniformHashDistribution)
+	{
+		useBinarySearch = true;
+	}
+
+	/* we only need hash functions for hash distributed tables */
+	if (partitionMethod == DISTRIBUTE_BY_HASH)
+	{
+		hashFunction = cacheEntry->hashFunction;
+	}
+
+	/*
+	 * Call FindShardInterval to find the corresponding shard interval for the
+	 * given partition value.
+	 */
+	shardInterval = FindShardInterval(partitionValue, sortedShardIntervalArray,
+									  shardCount, partitionMethod,
+									  shardIntervalCompareFunction, hashFunction,
+									  useBinarySearch);
+
+	return shardInterval;
+}
+
+
+/*
+ * PrunableExpressions builds a list of all prunable expressions in node,
+ * storing them in context->pruningInstances.
+ */
+static void
+PrunableExpressions(Node *node, ClauseWalkerContext *context)
+{
+	/*
+	 * Build initial list of prunable expressions.  As long as only,
+	 * implicitly or explicitly, ANDed expressions are found, this perform a
+	 * depth-first search.  When an ORed expression is found, the current
+	 * PruningInstance is added to context->pruningInstances (once for each
+	 * ORed expression), then the tree-traversal is continued without
+	 * recursing.  Once at the top-level again, we'll process all pending
+	 * expressions - that allows us to find all ANDed expressions, before
+	 * recursing into an ORed expression.
+	 */
+	PrunableExpressionsWalker(node, context);
+
+	/*
+	 * Process all pending instances.  While processing, new ones might be
+	 * added to the list, so don't use foreach().
+	 */
+	while (context->pendingInstances != NIL)
+	{
+		PendingPruningInstance *instance =
+			(PendingPruningInstance *) linitial(context->pendingInstances);
+		PruningInstance *oldPrune = instance->instance;
+		PruningInstance *newPrune = palloc0(sizeof(PruningInstance));
+
+		if (oldPrune)
+		{
+			memcpy(newPrune, oldPrune, sizeof(PruningInstance));
```

Comment:
This probably deserves some more explanation. e.g. implement the distributive property: (P AND (Q || R)) `oldPrune` is a partial pruning instance which contains `P` which is copied into two new pruning instances, one for `P AND Q` and one for `P AND R`.